### PR TITLE
Remove deactivated/domain removed users on profile deletion

### PR DESCRIPTION
**Pull request by minhaminha**
## Product Description

Previously, deactivated mobile users or webusers no longer part of a domain would block a profile from being deleted if their user_data was still foreign key'd to the profile. This PR ensures that those FKs are set to null when attempting to delete a user profile if deactivated/domain removed users are the only types of users still associated with the profile. 

## Technical Summary

The foreign key on_delete is changed from PROTECT to SET_NULL and additional validation steps are introduced to only delete a profile if there aren't active mobile workers or webusers belonging to the domain the profile belongs to. 

### Safety story

Tested locally and on staging. 
Data changes is limited only to the `user_data` object of deactivated/domain removed users associated with the profile being deleted. 

### Automated test coverage

Added test coverage.

### QA Plan

N/A

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by Jtang-1** [comment]
From what I understand, the issue is that profiles can't be deleted because they are still assigned to users who have been removed from the domain.

The proposed solution is to unassign the profile from the user during either:

1. User deactivation
2. User domain membership removal (while saving the profile for potential reinstatement)

I’m concerned that this approach of modifying data during deactivation or domain removal sets an unsustainable pattern. Is a solution that during profile deletion, allow deletion if it is not assigned to an active user of the domain (and then remove the profile from all users). I'm not sure how it should handle a user that has had its profile deleted should be restored (especially if the profile configured to be required) but i'm also not seeing how that's handled in this PR

**Comment by minhaminha** [comment]
> I’m concerned that this approach of modifying data during deactivation or domain removal sets an unsustainable pattern.

Can you clarify why its unsustainable? Is it because there might be in the future more ways to remove/deactivate users that won't take into account the need to remove the profile at the same time?

> Is a solution that during profile deletion, allow deletion if it is not assigned to an active user of the domain (and then remove the profile from all users). I'm not sure how it should handle a user that has had its profile deleted should be restored (especially if the profile configured to be required) but i'm also not seeing how that's handled in this PR

I chose to remove the profile on removal/deactivation mostly so that, at least for webusers, I can capture any existing profile_ids in its DomainRemovalRecord, to make restoring the profile on undo possible. I guess it would be possible to modify the records after it's been created/while attempting to remove the profile but I thought this was more efficient. I'm planning to address restoring the profile for commcare users in a separate ticket/PR because it doesn't already have a record object like that to reference when reactivating the user so I think it'll require a little more lift to implement.

**Comment by Jtang-1** [comment]
@minhaminha 

>  Is it because there might be in the future more ways to remove/deactivate users that won't take into account the need to remove the profile at the same time?

That is a possibility but no, that's not my concern

What I’m getting at is that altering user data during removal or deactivation feels out of scope for those actions. I don’t have this thought fully fleshed out - but there's a lot of data associated with users, so I'm imagining it would get messy if in the future, something else comes up and we decide that it's also okay to custom handle that during user deactivation/removal from domain. Not sure if that make sense, would like to get Ethan's thought on this. I could be completely off base

> I can capture any existing profile_ids in its DomainRemovalRecord, to make restoring the profile on undo possible. 

If we don't delete the profile upon user deactivation/removal, wouldn’t we avoid the need to store it in the first place? The profile would stay associated with the user data, and if the user is restored, the profile would automatically be reinstated, right?

**Comment by minhaminha** [comment]
Ohh ok I think I kind of understand where you're coming from with this but for what it's worth, when we remove a web user we do other things like removing the associated tableau user too so its not like its currently only scoped to modifying the user object itself. I guess this does create an opportunity for user data to get messed up if for example there's an error mid-removal or undo but at least for this change, the stakes are very low (if the profile is missing, you can assign it again) and this doesn't seem to be a very error-prone area (but I could wrong). Also curious to hear from another dev on this

> If we don't delete the profile upon user deactivation/removal, wouldn’t we avoid the need to store it in the first place? 

True! I was also think from a performance perspective, where if you removed profiles from deactivated/removed users when attempting to delete the profile you'd have to get all users associated with it at once (`CustomDataFieldsProfile` has a `user_ids_assigned` method that uses the mysterious `self.sqluserdata_set` to get associated user_ids which I can use to grab user objects, which is already two requests)

**Comment by Jtang-1** [comment]
I think the case with removing tableau user is different and makes more sense to do since, irrc, it’s mapped directly one to one.

The performance hit you described doesn’t sound too bad, particularly because I don’t think profile deletion happens very often.

File: corehq/apps/users/util.py [hunk 1]
```diff
@@ -441,6 +442,18 @@ def bulk_auto_deactivate_commcare_users(user_ids, domain):
...
+    # Remove profiles
+    user_objs = []
+    for user_id in user_ids:
+        user_objs.append(CommCareUser.get_by_user_id(user_id))
+    users = prime_user_data_caches(user_objs, domain)
+    for user in users:
+        user_data = user.get_user_data(domain)
+        if user_data.profile:
+            data = user_data.to_dict()
+            user_data.update(data, '')
```
Comment (esoergel) [inline corehq/apps/users/util.py:454]:
```suggestion
            user_data.profile_id = None
```
like here:
https://github.com/dimagi/commcare-hq/blob/d1619792493f653c4936aa27d98c1f7f52023a34/corehq/apps/users/tests/test_user_data.py#L124-L126

**Comment by esoergel** [comment]
The main trouble here is making it undo-able, right?  I don't have any issue removing the profile (and all user data) when removing from the domain.  That's all domain-associated data, that's conceptually part of the `DomainMembership`, even if it's stored elsewhere.  Removing a user from a domain _is_ removing their domain membership.  If this were all stored in SQL, I'd have `user_data` probably be FK'd to the domain membership and cascade on deletion.

That's actually a big part of the problem here, that we're mixing couch and SQL.  This isn't the first time this sort of issue has arisen - assigned locations are also stored on the domain membership, so what happens to those when a user is deleted?  Since it's not an actual FK reference, I'd wager that we keep a record of the ID around in this `DomainRemovalRecord`.  Then if you try to delete the location we probably don't actually find the ID reference in the old domain membership when warning about assigned users, nor when [unassigning those users](https://github.com/dimagi/commcare-hq/blob/d1619792493f653c4936aa27d98c1f7f52023a34/corehq/apps/locations/models.py#L438-L438).  This means that undoing that domain removal record will leave the user with location assignments that point to deleted locations.

So in CouchDB we have to be really explicit and careful when handling this kind of thing, and it's easy to accidentally forget to handle this sort of situation.  With SQL in contrast, we can define an [on_delete](https://github.com/dimagi/commcare-hq/blob/d1619792493f653c4936aa27d98c1f7f52023a34/corehq/apps/users/user_data.py#L212-L213) for the relationship.

Anyways, so I think the options are:

* Remove user data entirely on users when we remove them from a domain.  This is also what happens when you delete a mobile worker, along with various other irreversible actions (see the various warnings when you go to delete a MW).  I think that's my preference.
* Keep

**Comment by minhaminha** [comment]
Sorry @esoergel for the delay - completely missed that you made a lengthy comment here. 

I think I like your hybrid approach the best but I'm curious why you'd prefer the strict version. The difference between reactivating a mobile worker vs restoring a web user is that the latter can only be done (through the UI) I believe immediately after removing them from the domain so I think it make sense to have everything restored as it was.

**Comment by esoergel** [comment]
@minhaminha I don't know that I feel that strongly about it - the strict version is nice to just avoid a dangling reference, feels a bit cleaner.  But the hybrid is probably a better UX :shrug:

**Review by esoergel — APPROVED** [review]

File: corehq/apps/custom_data_fields/tests/test_profiles.py [hunk 2]
```diff
@@ -106,3 +112,37 @@ def test_users_assigned(self):
         self.assertItemsEqual([], self.profile3.user_ids_assigned())
         self.assertTrue(self.profile5.has_users_assigned)
         self.assertItemsEqual([user._id], self.profile5.user_ids_assigned())
+
+    def test_profile_wont_delete_if_mobile_user_is_active(self):
```
Comment (Jtang-1) [inline corehq/apps/custom_data_fields/tests/test_profiles.py:116]: ⭐

**Review by Jtang-1 — APPROVED** [review]
looks good! thank you for talking through my thoughts with me

**Pull request merged**